src/zkbob/ZkBobPool.sol

```diff
@@ -169,34 +183,39 @@ contract ZkBobPool is EIP1967Admin, Ownable, Parameters, ZkBobAccounting {
 
         if (txType == 0) {
             // Deposit
-            require(token_amount >= 0 && energy_amount == 0 && msg.value == 0, "ZkBobPool: incorrect deposit amounts");
+            require(token_amount >= 0 && energy_amount == 0, "ZkBobPool: incorrect deposit amounts");
             IERC20(token).safeTransferFrom(user, address(this), uint256(token_amount) * TOKEN_DENOMINATOR);
         } else if (txType == 1) {
             // Transfer
-            require(token_amount == 0 && energy_amount == 0 && msg.value == 0, "ZkBobPool: incorrect transfer amounts");
+            require(token_amount == 0 && energy_amount == 0, "ZkBobPool: incorrect transfer amounts");
         } else if (txType == 2) {
             // Withdraw
-            require(
-                token_amount <= 0 && energy_amount <= 0 && msg.value == _memo_native_amount() * NATIVE_DENOMINATOR,
-                "ZkBobPool: incorrect withdraw amounts"
-            );
+            require(token_amount <= 0 && energy_amount <= 0, "ZkBobPool: incorrect withdraw amounts");
+
+            uint256 native_amount = _memo_native_amount() * TOKEN_DENOMINATOR;
+            uint256 withdraw_amount = uint256(-token_amount) * TOKEN_DENOMINATOR;
+
+            if (native_amount > 0) {
+                ITokenSeller seller = tokenSeller;
+                if (address(seller) != address(0)) {
+                    withdraw_amount -= native_amount;
+                    IERC20(token).safeTransfer(address(seller), native_amount);
+                    seller.sellForETH(user, native_amount);
+                }
+            }
 
-            if (token_amount < 0) {
-                IERC20(token).safeTransfer(user, uint256(-token_amount) * TOKEN_DENOMINATOR);
+            if (withdraw_amount > 0) {
```

Comment:
Is it possible when `abs(withdraw_amount + amount actually exchanged for native tokens)` is less than `abs(token_amount)`?

Reply:
Yes, I added the logic to refund any unspent tokens to the user.